### PR TITLE
Major first-stage refactor. Faster compilation, method syntax works predictably, and semicolons!

### PR DESCRIPTION
**Pull request by dfellis**
Resolves \#98 and \#330

Currently only reproduces the current behavior with new ANTLR grammar, but the new grammar should allow new tests to confirm that it's fixed.

### CODE REVIEW

**Comment by dfellis** [comment]
This is probably going to be the main thing I work on this week.

Also the current draft of the ANTLR grammar looks correct to me, but it's also much slower than the existing grammar, so I may completely rework this again before it's landed -- I would like to *reduce* the amount of backtracking, not increase it, in the parser.

**Comment by dfellis** [comment]
Creating this comment to keep track of which test suites are working and which still aren't. Will just update this comment as they're resolved now.

- [x] 001_event_spec.sh
- [x] 002_printing.sh
- [ ] 003_math_spec.sh
- [x] 004_bitwise_spec.sh
- [x] 005_boolean_spec.sh
- [ ] 006_string_spec.sh
- [x] 007_comparator_spec.sh
- [x] 008_function_spec.sh
- [ ] 009_conditional_spec.sh
- [ ] 010_object_literal_spec.sh
- [ ] 011_array_spec.sh
- [ ] 012_hash_spec.sh
- [x] 013_generics_spec.sh
- [ ] 014_interface_spec.sh
- [ ] 015_maybe_result_either_spec.sh
- [x] 016_type_spec.sh
- [x] 017_custom_event_spec.sh
- [ ] 018_closure_spec.sh
- [x] 019_compiler_error_spec.sh
- [x] 020_module_constants_spec.sh
- [ ] 021_trig_spec.sh
- [x] 022_deps_spec.sh
- [ ] 023_http_spec.sh
- [ ] 024_clone_spec.sh
- [x] 025_runtime_error_spec.sh
- [ ] 026_datastore_spec.sh (currently hangs)
- [ ] 027_seq_spec.sh
- [ ] 028_tree_spec.sh
- [x] 029_err_printing_spec.sh
- [x] 030_cmd_spec.sh

**Comment by dfellis** [comment]
It's all passing locally, now!

Next steps:

- [ ] Add test cases for syntax this work is meant to unlock
- [ ] Make function return type determination lazy so we actually get the perf gain this is meant to provide
- [ ] Rework the website and docs to scatter semicolons across them :/

**Comment by dfellis** [comment]
Soo... I can't reproduce this error locally, but it appears to be consistent across the three environments. Trying to figure out why.

File: std/root.ln
Comment (dfellis) [inline std/root.ln:844]: Take a look at this. It seems trivial, but this kind of thing wasn't parsing correctly before!

**Comment by dfellis** [comment]
> Soo... I can't reproduce this error locally, but it appears to be consistent across the three environments. Trying to figure out why.

Pulled out an old laptop and rebuilt and tested on there and also can't reproduce the error here on that branch.

I'm going to temporarily turn off the xargs-based parallel execution and see if that helps, because I'm at a loss.

**Comment by dfellis** [comment]
This error is maddeningly consistent and nonsensical! The line numbers its pointing to don't line up with any assertions in the shellspec file!

**Comment by dfellis** [comment]
Discovered an issue with the new type inference code in this branch with code that uses an outer scope variable failing.

This *must* be resolved before this PR is landed. Putting this back into draft mode.

File: std/trig.ln
Comment (depombo) [inline std/trig.ln:41]: 👀

File: compiler/src/lntoamm/Statement.ts
Comment (depombo) [inline compiler/src/lntoamm/Statement.ts:95]: nit / not blocking this PR: what is "this" and what is a potential Redo? I actually have hit this error a few times

File: compiler/src/lntoamm/Microstatement.ts
Comment (depombo) [inline compiler/src/lntoamm/Microstatement.ts:1202]: this is awesome!!!

**Review by depombo — APPROVED** [review]

File: compiler/src/lntoamm/Statement.ts
Comment (dfellis) [reply compiler/src/lntoamm/Statement.ts:95]: I deleted everything and just marked all assignables pure all the time. So it needs to be completely redone and I thought it was self-evident that `return true` is not something you'd ever actually want from a function call.

We aren't yet using function/statement purity in the compiler, so it's not a huge deal, but if/when we want the compiler to automatically downgrade a `map` to a `mapLin` because you're mutating outer state, this will need to be working and correct.

I can update the TODO comment with all of this if you want?

Comment (dfellis) [reply compiler/src/lntoamm/Statement.ts:95]: Since it's not blocking, though, I can do it in a followup PR.

**Pull request merged**
